File: devicesetup/src/main/java/io/particle/android/sdk/devicesetup/ui/PermissionsFragment.java
package io.particle.android.sdk.devicesetup.ui;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.ActivityCompat.OnRequestPermissionsResultCallback;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AlertDialog.Builder;
import android.util.Log;

import java.util.Arrays;

import io.particle.android.sdk.devicesetup.R;
import io.particle.android.sdk.utils.WorkerFragment;
import io.particle.android.sdk.utils.ui.Ui;

import static android.content.pm.PackageManager.PERMISSION_GRANTED;


public class PermissionsFragment extends Fragment implements OnRequestPermissionsResultCallback {

    public interface Client {
        void onUserAllowedPermission(String permission);

        void onUserDeniedPermission(String permission);
    }


    public static final String TAG = WorkerFragment.buildFragmentTag(PermissionsFragment.class);


    public static <T extends FragmentActivity & Client> PermissionsFragment ensureAttached(
            T callbacksActivity) {
        PermissionsFragment frag = get(callbacksActivity);
        if (frag == null) {
            frag = new PermissionsFragment();
            WorkerFragment.addFragment(callbacksActivity, frag, TAG);
        }
        return frag;
    }

    public static <T extends FragmentActivity & Client> PermissionsFragment get(T callbacksActivity) {
        return Ui.findFrag(callbacksActivity, TAG);
    }

    public static boolean hasPermission(@NonNull Context ctx, @NonNull String permission) {
        int result = ContextCompat.checkSelfPermission(ctx, permission);
        return (result == PERMISSION_GRANTED);
    }

    public void ensurePermission(final @NonNull String permission) {
        Builder dialogBuilder = new AlertDialog.Builder(getActivity())
                .setCancelable(false);

        // FIXME: stop referring to these location permission-specific strings here,
        // try to retrieve them from the client
        if (ActivityCompat.checkSelfPermission(getActivity(), permission) != PERMISSION_GRANTED ||
                ActivityCompat.shouldShowRequestPermissionRationale(getActivity(), permission)) {
            dialogBuilder.setTitle(R.string.location_permission_dialog_title)
                    .setMessage(R.string.location_permission_dialog_text)
                    .setPositiveButton(R.string.got_it, (dialog, which) -> {
                        dialog.dismiss();
                        requestPermission(permission);
                    });
        } else {
            // user has explicitly denied this permission to setup.
            // show a simple dialog and bail out.
            dialogBuilder.setTitle(R.string.location_permission_denied_dialog_title)
                    .setMessage(R.string.location_permission_denied_dialog_text)
                    .setPositiveButton(R.string.settings, (dialog, which) -> {
                        dialog.dismiss();
                        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                        String pkgName = getActivity().getApplicationInfo().packageName;
                        intent.setData(Uri.parse("package:" + pkgName));
                        startActivity(intent);
                    })
                    .setNegativeButton(R.string.exit_setup, (dialog, which) -> {
                        Client client = (Client) getActivity();
                        client.onUserDeniedPermission(permission);
                    });
        }

        dialogBuilder.show();
    }

    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        Log.i(TAG, String.format("onRequestPermissionsResult(%d, %s, %s)",
                requestCode,
                Arrays.toString(permissions),
                Arrays.toString(grantResults)));
        if (requestCode != REQUEST_CODE) {
            Log.i(TAG, "Unrecognized request code: " + requestCode);
        }

        // we only ever deal with one permission at a time, so we can always safely grab the first
        // member of this array.
        String permission = permissions[0];
        if (hasPermission(getActivity(), permission)) {
            Client client = (Client) getActivity();
            client.onUserAllowedPermission(permission);
        } else {
            this.ensurePermission(permission);
        }
    }

    private void requestPermission(String permission) {
        ActivityCompat.requestPermissions(getActivity(), new String[]{permission}, REQUEST_CODE);
    }

    private static final int REQUEST_CODE = 128;

}

File: meshui/src/main/java/io/particle/mesh/ui/setup/PermissionsFragment.kt
package io.particle.mesh.ui.setup

import android.content.Intent
import android.net.Uri
import android.provider.Settings
import android.util.Log
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.app.ActivityCompat.OnRequestPermissionsResultCallback
import androidx.core.content.PermissionChecker.PERMISSION_GRANTED
import androidx.core.content.PermissionChecker.checkSelfPermission
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.fragment.app.commit
import io.particle.android.sdk.utils.appHasPermission
import io.particle.mesh.ui.R
import java.util.*

class PermissionsFragment : Fragment(), OnRequestPermissionsResultCallback {

    companion object {

        const val TAG = "FRAG_MESH_PERMISSIONS"


        fun <T> ensureAttached(callbacksActivity: T): PermissionsFragment
                where T : FragmentActivity, T : Client {
            var frag: PermissionsFragment? = get(callbacksActivity)
            if (frag == null) {
                frag = PermissionsFragment()
                callbacksActivity.supportFragmentManager.commit { add(frag, TAG) }
            }
            return frag
        }

        fun <T> get(callbacksActivity: T): PermissionsFragment?
                where T : FragmentActivity, T : Client {
            val supportFragMan = callbacksActivity.supportFragmentManager
            return supportFragMan.findFragmentByTag(TAG) as PermissionsFragment?
        }

        private val REQUEST_CODE = 128
    }


    interface Client {
        fun onUserAllowedPermission(permission: String)
        fun onUserDeniedPermission(permission: String)
    }


    fun ensurePermission(permission: String) {
        val aktivity = activity!!

        if (aktivity.appHasPermission(permission)) {
            return
        }

        val dialogBuilder = AlertDialog.Builder(aktivity).setCancelable(false)

        // FIXME: stop referring to these location permission-specific strings here,
        // make them part of the client interface
        if (shouldShowRequestPermissionRationale(permission)) {
            dialogBuilder.setTitle(R.string.mesh_location_permission_dialog_title)
                .setMessage(R.string.mesh_location_permission_dialog_text)
                .setPositiveButton(R.string.mesh_got_it) { dialog, _ ->
                    dialog.dismiss()
                    requestPermission(permission)
                }
        } else {
            // user has explicitly denied this permission to setup.
            // show a simple dialog and bail out.
            dialogBuilder.setTitle(R.string.mesh_location_permission_denied_dialog_title)
                .setMessage(R.string.mesh_location_permission_denied_dialog_text)
                .setPositiveButton("Settings") { dialog, _ ->
                    dialog.dismiss()
                    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
                    val pkgName = aktivity.applicationInfo.packageName
                    intent.data = Uri.parse("package:$pkgName")
                    startActivity(intent)
                }
                .setNegativeButton("Exit") { _, _ ->
                    val client = activity as Client?
                    client!!.onUserDeniedPermission(permission)
                }
        }

        dialogBuilder.show()
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        val msg = "onRequestPermissionsResult($requestCode, ${Arrays.toString(permissions)}," +
                "${Arrays.toString(grantResults)})"
        Log.i(TAG, msg)

        if (requestCode != REQUEST_CODE) {
            Log.i(TAG, "Unrecognized request code: $requestCode")
        }

        // we only ever deal with one permission at a time, so we can always safely grab the first
        // member of this array.
        val permission = permissions[0]
        val client = activity!! as Client
        if (activity!!.appHasPermission(permissions[0])) {
            client.onUserAllowedPermission(permission)
        } else {
            client.onUserDeniedPermission(permission)
        }
    }

    private fun requestPermission(permission: String) {
        requestPermissions(arrayOf(permission), REQUEST_CODE)
    }

}
